### PR TITLE
Oauth

### PR DESCRIPTION
**Pull request by irenenikk**
Add oauth support
Add organizations (new command and prompt on login)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -181,6 +182,9 @@
                 <artifactId>cobertura-maven-plugin</artifactId>
                 <version>2.7</version>
                 <configuration>
+                    <check>
```
Comment (nygrenh) [inline pom.xml]: Is this necessary?

File: src/main/java/fi/helsinki/cs/tmc/cli/Application.java [hunk 2]
```diff
@@ -20,12 +20,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Date;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/Application.java]: I'd prefer not to use star imports.

File: src/main/java/fi/helsinki/cs/tmc/cli/Application.java [hunk 3]
```diff
@@ -176,6 +172,20 @@ public void run(String[] args) {
     }
 
     public static void main(String[] args) {
+        while(true) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/Application.java]: Do we want to have your debugging loop as the main entry point for all of our users?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Account.java [hunk 4]
```diff
@@ -1,35 +1,53 @@
...
+import fi.helsinki.cs.tmc.core.domain.OauthCredentials;
+import fi.helsinki.cs.tmc.core.domain.Organization;
+
 /**
  * This object stores all login info.
  */
 public class Account {
 
-    static Account NULL_ACCOUNT = new Account(null, null, null);
+    static Account NULL_ACCOUNT = new Account(null, null, null, null);
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Account.java]: Null null null null null null null null null null null... :camel:

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 5]
```diff
@@ -49,32 +51,40 @@ public void setWorkDir(WorkDir workDir) {
 
     @Override
     public String getServerAddress() {
-        return account.getServerAddress();
+        if (account.getServerAddress().isPresent()) {
+            return account.getServerAddress().get();
+        }
+        return "https://tmc.mooc.fi";
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: Maybe we should extract this into a constant?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 6]
```diff
@@ -92,7 +102,7 @@ public String getFormattedUserData() {
         if (!userDataExists()) {
             return "";
         }
-        return getUsername() + ":" + this.getPassword();
+        return getUsername().get() + ":" + this.getPassword().get();
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: What's the purpose of this? Isn't the password supposed to be almost always absent in the oauth branch?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 7]
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
     }
 
     @Override
-    public void setCourse(Course course) {}
+    public String hostProgramName() {
+        // which command line is used
+        return null;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: I think this should be implemented.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 8]
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
...
+    }
+
+    @Override
+    public String hostProgramVersion() {
+        // which command line is used
+        return null;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: You probably want to implement this too

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 9]
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
...
+    }
 
     @Override
-    public void setConfigRoot(Path path) {}
+    public boolean getSendDiagnostics() {
+        return false;
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: This should be an user-controlled setting. If it's always false, we won't ever get crash reports.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 10]
```diff
@@ -116,8 +126,58 @@ public Path getConfigRoot() {
...
+        account.setOrganization(organization);
+    }
+
+    @Override
+    public void setCourse(Course course) {
+        account.setCurrentCourse(Optional.of(course));
+    }
+
+    @Override
+    public void setConfigRoot(Path path) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: What is `setConfigRoot`? :confused:

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java [hunk 11]
```diff
@@ -142,7 +142,6 @@ private static boolean requireConfigDirectory(Path configRoot) {
             try {
                 Files.createDirectories(configRoot);
             } catch (Exception e) {
-                //TODO print error to user
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java]: Why did you remove this TODO? I don't think logger.error prints a message to the user or does it?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java [hunk 12]
```diff
@@ -156,7 +155,6 @@ private static AccountList getHolderFromJson(Path file) {
         try {
             reader = Files.newBufferedReader(file, Charset.forName("UTF-8"));
         } catch (IOException e) {
-            //TODO print error to user
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java]: See above.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java [hunk 13]
```diff
@@ -169,7 +167,6 @@ private static boolean saveHolderToJson(AccountList holder, Path file) {
         try {
             Files.write(file, json);
         } catch (IOException e) {
-            //TODO print error to user
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java]: See the above of above :stuck_out_tongue:

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java [hunk 14]
```diff
@@ -216,34 +233,46 @@ private static void handleTmcExceptions(CliContext ctx, Exception exception) {
...
+
         if (cause != null && cause.getCause() instanceof UnknownHostException) {
             logger.error("No internet connection");
             io.errorln("You have no internet connection.");
             return;
         }
         logger.error("Command failed in tmc-core", exception);
         //TODO we seem to write twice error message; here and in the commands.
-        io.errorln("Command failed in tmc-core, check tmc-cli.log file for more info");
+        io.errorln("Command failed, check tmc-cli.log file for more info");
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java:269]: Yes :+1:. The tmc-core part was really confusing to the user.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java [hunk 15]
```diff
@@ -216,34 +233,46 @@ private static void handleTmcExceptions(CliContext ctx, Exception exception) {
...
     }
 
     private static boolean isAuthenticationError(Exception exception) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/TmcUtil.java:272]: Could be simplified to `return exception.getCause() instanceof OAuthProblemException`.

File: src/main/java/fi/helsinki/cs/tmc/cli/utils/OptionalToGoptional.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package fi.helsinki.cs.tmc.cli.utils;
+
+import com.google.common.base.Optional;
+
+public class OptionalToGoptional {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/utils/OptionalToGoptional.java:5]: :camel:

File: src/test/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIoTest.java [hunk 17]
```diff
@@ -53,4 +57,15 @@ public void loadingFromFileWorks() {
...
     }
+
+    @Test
+    public void abortingCourseCreationWorks() {
+        CourseInfoIo.save(this.course, this.courseFile);
+        CourseInfo loadedInfo = CourseInfoIo.load(this.courseFile);
+        Assert.assertNotNull(loadedInfo);
+
+        CourseInfoIo.abortCreatingCourse(new Course("test-course"), Paths.get(tempDir));
+        File courseJson = Paths.get(tempDir).resolve(".tmc.json").toFile();
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIoTest.java]: Avoid using `java.io.File` in new code and prefer the better alternatives in `java.nio`.

For example, this could be rewritten as:
```java
Path courseJson = Paths.get(tempDir).resolve(".tmc.json");
assertTrue(Files.exist(courseJson));
```

File: src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java [hunk 18]
```diff
@@ -23,7 +9,7 @@
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
 
-import java.awt.Desktop;
+import java.awt.*;
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java]: Avoid star imports

File: src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java [hunk 19]
```diff
@@ -33,6 +19,16 @@
 import java.nio.file.Paths;
 import java.util.List;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.internal.matchers.StringContains.containsString;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.*;
```
Comment (nygrenh) [inline src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java]: Avoid star imports

File: src/main/java/fi/helsinki/cs/tmc/cli/command/OrganizationCommand.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        organizations.stream()
+                .sorted(Comparator.comparing(Organization::getName))
+                .forEach(o -> {
+                    if (o.isPinned()) {
+                        pinned.add(o);
+                    } else {
+                        others.add(o);
+                    }
+                });
+        System.out.println();
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/command/OrganizationCommand.java]: Some parts of the code use `System.println` for printing and other parts `io.println`.

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (irenenikk) [reply pom.xml]: According to IntelliJ, yes. Without the check-tag, the configuration tag is shown in red.

File: src/main/java/fi/helsinki/cs/tmc/cli/Application.java (hunk 2 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/Application.java]: :palm_tree:

File: src/main/java/fi/helsinki/cs/tmc/cli/Application.java (hunk 3 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/Application.java]: :dolphin:

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java (hunk 5 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: :tulip:

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Account.java (hunk 4 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/backend/Account.java]: The Null Account is a part of the original software. It is used when the method `setAccount(Account a)` of `Settings` is called with `a` being null, so that the getters wouldn't throw the `NullPointerException`. `setAccount` was added by the original team, and I did not find any use of it where `account` would seem to be null, so maybe deleting it would indeed be a good idea.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java (hunk 6 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: It's true that password is actually always `null` in this branch. However to me it is unclear where the method is used in general (it is a part of the Settings interface) , so I thought that the point of the method is to get the password as it returned it before.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java (hunk 9 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: There is no setter for this method, so how is the behaviour determined?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java (hunk 10 above)
Comment (irenenikk) [reply src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: This has been discussed earlier and apparently one shouldn't be able to change the config root, so the method is pointless (NOP in tmc netbeans as well). It should be removed from the API.

File: src/test/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIoTest.java (hunk 17 above)
Comment (irenenikk) [reply src/test/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIoTest.java]: :whale:

File: src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java (hunk 18 above)
Comment (irenenikk) [reply src/test/java/fi/helsinki/cs/tmc/cli/io/ExternalUtilTest.java]: It seems like IntelliJ offers star imports, I will stop using the Alt+Enter auto import functionality.

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIo.java [hunk 21]
```diff
@@ -62,4 +62,17 @@ public static void createNewCourse(Course course, Account account, Path parentDi
         info.setExercises(course.getExercises());
         CourseInfoIo.save(info, configFile);
     }
+
+    public static void abortCreatingCourse(Course course, Path parentDir) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/CourseInfoIo.java]: Bit misleading name

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java [hunk 22]
```diff
@@ -49,32 +48,37 @@ public void setWorkDir(WorkDir workDir) {
 
     @Override
     public String getServerAddress() {
-        return account.getServerAddress();
+            return account.getServerAddress();
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/Settings.java]: Why did you add this indentation?

File: src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java [hunk 23]
```diff
@@ -67,6 +70,24 @@ public static boolean delete() {
         return true;
     }
 
+    public static void saveCurrentSettingsToAccountList(Settings settings) {
+        AccountList list = loadAccountList();
+        Set<Account> deletables = new HashSet<>();
+        for (Iterator<Account> i = list.iterator(); i.hasNext(); ) {
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/backend/SettingsIo.java]: For each?

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: src/main/java/fi/helsinki/cs/tmc/cli/shared/CourseFinder.java [hunk 24]
```diff
@@ -88,7 +88,7 @@ private boolean handleMultipleMatchingCourses(Map<Account, Course> matches) {
                     "Download course from "
                             + entryAccount.getServerAddress()
                             + " with '"
-                            + entryAccount.getUsername()
+                            + entryAccount.getUsername().get()
```
Comment (nygrenh) [inline src/main/java/fi/helsinki/cs/tmc/cli/shared/CourseFinder.java:95]: Nullpointerexception?

**Review by nygrenh — CHANGES_REQUESTED** [review]

**Review by nygrenh — APPROVED** [review]
